azavea/pwd-river-dispatches
# Fallback Support

## Overview

Refactors the app organization a bit, to add the following discipline to the various metrics:

```js
metric: {
  code: '00010',
  units: '°F',
  transform: function(c) {
    var f = c * 9 / 5 + 32;
    return Math.round(f * 10) / 10;
  },
  fallback: [
    42, 32, 42, 51, 61, 71, 80, 90, 80, 71, 61, 51
  ]
}
```

Each metric has the above properties, which are described below:

  * `code`: For the API
  * `units`: For the UI
  * `transform`: A function that converts API values to UI values
  * `fallback`: An array of 12 values that has the fallback value for that month of the year.

`fallback` values are used when the API returns a non-`200 OK` response. The listener now uses a factory pattern to create one specifically for the metric to handle.

Connects #13 
Connects #14 

### Demo

![image](https://user-images.githubusercontent.com/1430060/27357211-46b84caa-55df-11e7-9384-79a1422b9603.png)

### Notes

Turbidity has not been transformed into the appropriate values. We'll need more feedback before we can do that. #16 
If WiFi is disconnected, we see nothing instead of the fallback values. Need to figure out how to handle browser-level errors like that. #15 

This may be getting just to the size to warrant splitting out into separate files. I haven't done it this time, but if this file grows any more we'll have to do it.

## Testing Instructions

 * Run `server`
 * Check the Temperature section (top right, Orange). Ensure you see values in ℉.
 * Change the `BASE_URL` to have the wrong path:

```diff
diff --git a/usgs-details.html b/usgs-details.html
index 4cebf71..919e835 100644
--- a/usgs-details.html
+++ b/usgs-details.html
@@ -31,7 +31,7 @@
     </div>
     <script type="text/javascript">
         var SITE_ID = '01474500';
-        var BASE_URL = 'https://waterservices.usgs.gov/nwis/iv/';
+        var BASE_URL = 'https://waterservices.usgs.gov/nwis/iv/notTheRightPath';
         var REQUEST_URL = BASE_URL + '?format=json' + '&sites=' + SITE_ID;
 
         var METRICS = {
```

 * Ensure that the app still works, and falls back to predefined values. Check the console for messages.
